File: go/orm/migrations/schema_snapshot.go
package migrations

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
)

// TableColumn is one column of a TableSchema, in storage terms.
type TableColumn struct {
	Name        string
	StorageType string
	Nullable    bool
	Key         bool
	Generated   bool
}

// IndexPart is one column of a TableIndex, in index order.
type IndexPart struct {
	ColumnName string
	Descending bool
}

// TableIndex is one index of a TableSchema.
type TableIndex struct {
	Name    string
	Columns []IndexPart
	Unique  bool
}

// TableSchema is a table's shape, provider-agnostic but in storage types
// (what CREATE TABLE emits) — the per-table schema snapshot, format v2
// (ADR-0013 add.3, ADR-0017). Both snapshot producers build this: the
// metadata exporter (a later `snapshot` command) and a (later) shadow
// replayer's introspection — which is what makes the two comparable.
type TableSchema struct {
	Name    string
	Columns []TableColumn
	Indexes []TableIndex
}

// FromMap is the current model rendered as a schema (the metadata-side producer).
func FromMap(m *core.EntityMap, dialect core.Dialect) *TableSchema {
	columns := make([]TableColumn, len(m.Properties))
	for i, p := range m.Properties {
		columns[i] = TableColumn{
			Name: p.ColumnName, StorageType: dialect.StorageType(p), Nullable: p.IsNullable,
			Key: p.IsKey, Generated: p.IsGenerated,
		}
	}
	indexes := make([]TableIndex, len(m.Indexes))
	for i, index := range m.Indexes {
		parts := make([]IndexPart, len(index.Columns))
		for j, c := range index.Columns {
			parts[j] = IndexPart{ColumnName: c.ColumnName, Descending: c.Descending}
		}
		indexes[i] = TableIndex{Name: index.Name, Columns: parts, Unique: index.Unique}
	}
	return &TableSchema{Name: m.RelationName, Columns: columns, Indexes: indexes}
}

// Export renders map's current shape as a format-v2 table snapshot document (see ExportSchema).
func Export(m *core.EntityMap, dialect core.Dialect, asOfVersion int64, generatedAt time.Time) string {
	return ExportSchema(FromMap(m, dialect), asOfVersion, generatedAt)
}

// ExportSchema is the format-v2 document for schema (ADR-0017): object,
// asOfVersion, generatedAt, then columns and indexes name-sorted (ordinal byte
// order) for determinism — byte-identical to the C# reference's export.
func ExportSchema(schema *TableSchema, asOfVersion int64, generatedAt time.Time) string {
	columns := append([]TableColumn(nil), schema.Columns...)
	sort.Slice(columns, func(i, j int) bool { return columns[i].Name < columns[j].Name })
	columnsJSON := make([]any, len(columns))
	for i, c := range columns {
		obj := core.JSONObject{}
		obj = obj.Set("column", c.Name)
		obj = obj.Set("type", c.StorageType)
		obj = obj.Set("nullable", c.Nullable)
		if c.Key {
			obj = obj.Set("key", true)
		}
		if c.Generated {
			obj = obj.Set("generated", true)
		}
		columnsJSON[i] = obj
	}

	indexes := append([]TableIndex(nil), schema.Indexes...)
	sort.Slice(indexes, func(i, j int) bool { return indexes[i].Name < indexes[j].Name })
	indexesJSON := make([]any, len(indexes))
	for i, index := range indexes {
		parts := make([]any, len(index.Columns))
		for j, part := range index.Columns {
			direction := "asc"
			if part.Descending {
				direction = "desc"
			}
			partObj := core.JSONObject{}
			partObj = partObj.Set("column", part.ColumnName)
			partObj = partObj.Set("direction", direction)
			parts[j] = partObj
		}
		obj := core.JSONObject{}
		obj = obj.Set("name", index.Name)
		obj = obj.Set("columns", parts)
		if index.Unique {
			obj = obj.Set("unique", true)
		}
		indexesJSON[i] = obj
	}

	doc := core.JSONObject{}
	doc = doc.Set("object", schema.Name)
	doc = doc.Set("asOfVersion", asOfVersion)
	doc = doc.Set("generatedAt", core.FormatUTC(generatedAt))
	doc = doc.Set("columns", columnsJSON)
	doc = doc.Set("indexes", indexesJSON)
	return core.CanonicalJSON(doc)
}

// ExportDDL is the DDL-shaped snapshot for view-, materialized-view-, and
// procedure-backed objects (ADR-0017 add.1): those self-reflect from their
// defining SQL, so their history is compared by (normalized) DDL, not columns.
func ExportDDL(objectName, kind, ddl string, asOfVersion int64, generatedAt time.Time) string {
	doc := core.JSONObject{}
	doc = doc.Set("object", objectName)
	doc = doc.Set("kind", kind)
	doc = doc.Set("asOfVersion", asOfVersion)
	doc = doc.Set("generatedAt", core.FormatUTC(generatedAt))
	doc = doc.Set("ddl", NormalizeDDL(ddl))
	return core.CanonicalJSON(doc)
}

type tableSnapshotColumnJSON struct {
	Column    string `json:"column"`
	Type      string `json:"type"`
	Nullable  bool   `json:"nullable"`
	Key       bool   `json:"key"`
	Generated bool   `json:"generated"`
}

type tableSnapshotIndexJSON struct {
	Name    string `json:"name"`
	Columns []struct {
		Column    string `json:"column"`
		Direction string `json:"direction"`
	} `json:"columns"`
	Unique bool `json:"unique"`
}

type tableSnapshotJSON struct {
	Object      string                    `json:"object"`
	AsOfVersion int64                     `json:"asOfVersion"`
	Columns     []tableSnapshotColumnJSON `json:"columns"`
	Indexes     []tableSnapshotIndexJSON  `json:"indexes"`
}

// ParseTable reads a format-v2 table snapshot document.
func ParseTable(data []byte) (*TableSchema, int64, error) {
	var doc tableSnapshotJSON
	if err := json.Unmarshal(data, &doc); err != nil {
		return nil, 0, fmt.Errorf("parse table snapshot: %w", err)
	}
	columns := make([]TableColumn, len(doc.Columns))
	for i, c := range doc.Columns {
		columns[i] = TableColumn{Name: c.Column, StorageType: c.Type, Nullable: c.Nullable, Key: c.Key, Generated: c.Generated}
	}
	indexes := make([]TableIndex, len(doc.Indexes))
	for i, index := range doc.Indexes {
		parts := make([]IndexPart, len(index.Columns))
		for j, part := range index.Columns {
			parts[j] = IndexPart{ColumnName: part.Column, Descending: part.Direction == "desc"}
		}
		indexes[i] = TableIndex{Name: index.Name, Columns: parts, Unique: index.Unique}
	}
	return &TableSchema{Name: doc.Object, Columns: columns, Indexes: indexes}, doc.AsOfVersion, nil
}

// ParseDDL reads a DDL-shaped snapshot document.
func ParseDDL(data []byte) (object, kind, ddl string, asOfVersion int64, err error) {
	var doc struct {
		Object      string `json:"object"`
		Kind        string `json:"kind"`
		AsOfVersion int64  `json:"asOfVersion"`
		DDL         string `json:"ddl"`
	}
	if err = json.Unmarshal(data, &doc); err != nil {
		return "", "", "", 0, fmt.Errorf("parse ddl snapshot: %w", err)
	}
	return doc.Object, doc.Kind, doc.DDL, doc.AsOfVersion, nil
}

var viewDDLPattern = regexp.MustCompile(
	`(?i)^create\s+(materialized\s+)?view\s+(if\s+not\s+exists\s+)?(\S+)\s+as\s+(.+)$`)

// NormalizeDDL is canonical, comparable, still-executable DDL: whitespace
// collapses (layout is not schema — mirrors the reference's Split(null,
// RemoveEmptyEntries) + Join(" ")), and a view create's prefix canonicalizes
// to lowercase "create [materialized] view <name> as" — databases rewrite
// that prefix when storing it (SQLite drops IF NOT EXISTS and recases CREATE
// VIEW), so the rendered and the introspected form must meet in the middle.
// The name and body are compared as written.
func NormalizeDDL(sql string) string {
	collapsed := strings.Join(strings.Fields(sql), " ")
	match := viewDDLPattern.FindStringSubmatch(collapsed)
	if match == nil {
		return collapsed
	}
	materialized := ""
	if strings.TrimSpace(match[1]) != "" {
		materialized = "materialized "
	}
	return "create " + materialized + "view " + match[3] + " as " + match[4]
}

// --- directory-based snapshot lookup (ADR-0017/0018) ------------------------
//
// The diff generator (strictly below the version it regenerates) and the
// shadow replayer's trusted-baseline restore (at or below a --from version,
// orm/sqlite/shadow.go) both need "the latest snapshot in one object's
// directory meeting a version bound". One scan-and-parse loop serves both
// (CODING-STANDARD §8).

// snapshotFilesIn lists dir's V*.schema.json files (unsorted; callers pick by version).
func snapshotFilesIn(dir string) []string {
	files, _ := filepath.Glob(filepath.Join(dir, "V*.schema.json"))
	return files
}

// latestTableSnapshot scans dir's table snapshots for the one with the
// highest asOfVersion for which accept holds.
func latestTableSnapshot(dir string, accept func(asOfVersion int64) bool) (schema *TableSchema, atVersion int64, ok bool) {
	atVersion = -1
	for _, file := range snapshotFilesIn(dir) {
		data, err := os.ReadFile(file)
		if err != nil {
			continue
		}
		candidate, asOfVersion, err := ParseTable(data)
		if err != nil {
			continue
		}
		if accept(asOfVersion) && asOfVersion > atVersion {
			atVersion, schema, ok = asOfVersion, candidate, true
		}
	}
	return schema, atVersion, ok
}

// latestDDLSnapshot scans dir's view/materialized-view DDL snapshots for the
// one with the highest asOfVersion for which accept holds.
func latestDDLSnapshot(dir string, accept func(asOfVersion int64) bool) (object, ddl string, atVersion int64, ok bool) {
	atVersion = -1
	for _, file := range snapshotFilesIn(dir) {
		data, err := os.ReadFile(file)
		if err != nil {
			continue
		}
		parsedObject, _, parsedDDL, asOfVersion, err := ParseDDL(data)
		if err != nil {
			continue
		}
		if accept(asOfVersion) && asOfVersion > atVersion {
			object, ddl, atVersion, ok = parsedObject, parsedDDL, asOfVersion, true
		}
	}
	return object, ddl, atVersion, ok
}

// LatestTableSnapshotAtOrBefore is dir's latest table snapshot with
// asOfVersion <= version — the shadow replayer's trusted-baseline read: a
// --from version is trusted, never verified below it.
func LatestTableSnapshotAtOrBefore(dir string, version int64) (schema *TableSchema, atVersion int64, ok bool) {
	return latestTableSnapshot(dir, func(v int64) bool { return v <= version })
}

// LatestDDLSnapshotAtOrBefore is dir's latest view/materialized-view DDL
// snapshot with asOfVersion <= version (see LatestTableSnapshotAtOrBefore).
func LatestDDLSnapshotAtOrBefore(dir string, version int64) (object, ddl string, atVersion int64, ok bool) {
	return latestDDLSnapshot(dir, func(v int64) bool { return v <= version })
}

File: php/src/Migrations/SchemaSnapshot.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Migrations;

use DateTimeImmutable;
use SimpleOrm\Dialect\Dialect;
use SimpleOrm\Json\CanonicalWriter;
use SimpleOrm\Metadata\EntityIndex;
use SimpleOrm\Metadata\EntityMap;
use SimpleOrm\Metadata\IndexColumn;
use SimpleOrm\Metadata\PropertyMap;
use SimpleOrm\Types\Iso8601;

/**
 * The per-table schema snapshot (ADR-0013 add.3, format v2 per ADR-0017): a
 * generated artifact (`Migrations/Table/<Object>/V000N.schema.json`) recording
 * the table's shape as of a version, stamped with generation time. Storage
 * types make snapshots directly DDL-usable — derived downs, trusted baseline
 * rebuilds — and comparable with database introspection. Tables only — views,
 * materialized views, and procedures self-reflect by DDL (`exportDdl()`).
 *
 * Every document is written through `Json\CanonicalWriter` (never
 * `json_encode`) so the byte layout matches the C# `System.Text.Json` reference
 * exactly — the conformance suite compares these documents structurally, but
 * a byte-identical writer is what makes that possible across languages.
 */
final class SchemaSnapshot
{
    private function __construct()
    {
    }

    /** The current model rendered as a schema (the metadata-side producer). */
    public static function fromMap(EntityMap $map, Dialect $dialect): TableSchema
    {
        $columns = array_map(
            static fn (PropertyMap $property): TableSchemaColumn => new TableSchemaColumn(
                $property->columnName,
                $dialect->storageType($property),
                $property->nullable,
                $property->key,
                $property->generated,
            ),
            $map->properties,
        );

        $indexes = array_map(
            static fn (EntityIndex $index): TableSchemaIndex => new TableSchemaIndex(
                $index->name,
                array_map(
                    static fn (IndexColumn $column): TableSchemaIndexPart => new TableSchemaIndexPart(
                        $column->columnName,
                        $column->descending,
                    ),
                    $index->columns,
                ),
                $index->unique,
            ),
            $map->indexes,
        );

        /** @var string $relationName table-kind maps always carry a relation name */
        $relationName = $map->relationName;

        return new TableSchema($relationName, $columns, $indexes);
    }

    public static function export(TableSchema $schema, int $asOfVersion, DateTimeImmutable $generatedAt): string
    {
        $columns = $schema->columns;
        usort($columns, static fn (TableSchemaColumn $a, TableSchemaColumn $b): int => strcmp($a->name, $b->name));
        $indexes = $schema->indexes;
        usort($indexes, static fn (TableSchemaIndex $a, TableSchemaIndex $b): int => strcmp($a->name, $b->name));

        return CanonicalWriter::write([
            'object' => $schema->name,
            'asOfVersion' => $asOfVersion,
            'generatedAt' => self::formatGeneratedAt($generatedAt),
            'columns' => array_map(self::columnDocument(...), $columns),
            'indexes' => array_map(self::indexDocument(...), $indexes),
        ]);
    }

    /**
     * The DDL-shaped snapshot for view-, materialized-view-, and
     * procedure-backed objects: those self-reflect from their defining SQL, so
     * their history is compared by DDL, not by columns. The DDL is
     * whitespace-normalized — layout changes in the source SQL are not schema
     * changes — and stays executable.
     */
    public static function exportDdl(string $objectName, string $kind, string $ddl, int $asOfVersion, DateTimeImmutable $generatedAt): string
    {
        return CanonicalWriter::write([
            'object' => $objectName,
            'kind' => $kind,
            'asOfVersion' => $asOfVersion,
            'generatedAt' => self::formatGeneratedAt($generatedAt),
            'ddl' => self::normalizeDdl($ddl),
        ]);
    }

    public static function parseDdl(string $json): ParsedDdlSnapshot
    {
        /** @var array{object: string, kind: string, ddl: string, asOfVersion: int} $data */
        $data = json_decode($json, true, flags: JSON_THROW_ON_ERROR);

        return new ParsedDdlSnapshot($data['object'], $data['kind'], $data['ddl'], (int) $data['asOfVersion']);
    }

    public static function parse(string $json): ParsedTableSnapshot
    {
        /** @var array{object: string, asOfVersion: int, columns: list<array<string, mixed>>, indexes: list<array<string, mixed>>} $data */
        $data = json_decode($json, true, flags: JSON_THROW_ON_ERROR);

        $columns = array_map(
            static fn (array $c): TableSchemaColumn => new TableSchemaColumn(
                (string) $c['column'],
                (string) $c['type'],
                (bool) $c['nullable'],
                (bool) ($c['key'] ?? false),
                (bool) ($c['generated'] ?? false),
            ),
            $data['columns'],
        );

        $indexes = array_map(
            static fn (array $i): TableSchemaIndex => new TableSchemaIndex(
                (string) $i['name'],
                array_map(
                    static fn (array $p): TableSchemaIndexPart => new TableSchemaIndexPart(
                        (string) $p['column'],
                        ($p['direction'] ?? 'asc') === 'desc',
                    ),
                    $i['columns'],
                ),
                (bool) ($i['unique'] ?? false),
            ),
            $data['indexes'],
        );

        return new ParsedTableSnapshot(new TableSchema($data['object'], $columns, $indexes), (int) $data['asOfVersion']);
    }

    /**
     * Canonical, comparable, still-executable DDL: whitespace collapses
     * (layout is not schema), and a view create's prefix is canonicalized to
     * lowercase `create [materialized] view <name> as` — databases rewrite
     * that prefix when storing it (SQLite drops `IF NOT EXISTS` and recases
     * `CREATE VIEW`), so the rendered and the introspected form must meet in
     * the middle. The body is compared as written.
     */
    public static function normalizeDdl(string $sql): string
    {
        $words = preg_split('/\s+/', trim($sql), -1, PREG_SPLIT_NO_EMPTY);
        $collapsed = implode(' ', $words === false ? [] : $words);

        $pattern = '/^create\s+(?<mat>materialized\s+)?view\s+(if\s+not\s+exists\s+)?(?<name>\S+)\s+as\s+(?<body>.+)$/i';
        if (preg_match($pattern, $collapsed, $matches) === 1) {
            $materialized = !empty($matches['mat']) ? 'materialized ' : '';

            return "create {$materialized}view {$matches['name']} as {$matches['body']}";
        }

        return $collapsed;
    }

    /** @return array<string, mixed> */
    private static function columnDocument(TableSchemaColumn $column): array
    {
        $document = ['column' => $column->name, 'type' => $column->storageType, 'nullable' => $column->nullable];
        if ($column->key) {
            $document['key'] = true;
        }

        if ($column->generated) {
            $document['generated'] = true;
        }

        return $document;
    }

    /** @return array<string, mixed> */
    private static function indexDocument(TableSchemaIndex $index): array
    {
        $document = [
            'name' => $index->name,
            'columns' => array_map(
                static fn (TableSchemaIndexPart $part): array => [
                    'column' => $part->columnName,
                    'direction' => $part->descending ? 'desc' : 'asc',
                ],
                $index->columns,
            ),
        ];
        if ($index->unique) {
            $document['unique'] = true;
        }

        return $document;
    }

    private static function formatGeneratedAt(DateTimeImmutable $generatedAt): string
    {
        return Iso8601::format($generatedAt);
    }
}
